# This patch adds support for the `cause` property of Error objects in the concordance library,
# ensuring that when errors are described, their causes are also included.
# See https://github.com/concordancejs/concordance/pull/91 for more details.
diff --git a/lib/complexValues/error.js b/lib/complexValues/error.js
index bc96e7e54d4be5d2d6362c09479bbbba39959405..cea8599e2723f837d1a967a1c1a2746509be32bb 100644
--- a/lib/complexValues/error.js
+++ b/lib/complexValues/error.js
@@ -16,6 +16,7 @@ function describe (props) {
     name: error.name,
     messageIsEnumerable: isEnumerable(error, 'message'),
     message: error.message,
+    cause: error.cause,
   }, props))
 }
 exports.describe = describe
@@ -81,6 +82,7 @@ class DescribedErrorValue extends object.DescribedMixin(ErrorValue) {
     this.nameIsEnumerable = props.nameIsEnumerable
     this.messageIsEnumerable = props.messageIsEnumerable
     this.message = props.message
+    this.cause = props.cause
   }

   createPropertyRecursor () {
@@ -88,6 +90,7 @@ class DescribedErrorValue extends object.DescribedMixin(ErrorValue) {

     let skipName = this.nameIsEnumerable
     let emitMessage = !this.messageIsEnumerable
+    let emitCause = this.cause !== undefined

     let size = recursor.size
     if (skipName && size > 0) {
@@ -96,6 +99,9 @@ class DescribedErrorValue extends object.DescribedMixin(ErrorValue) {
     if (emitMessage) {
       size += 1
     }
+    if (emitCause) {
+      size += 1
+    }

     if (size === 0) return NOOP_RECURSOR

@@ -117,6 +123,11 @@ class DescribedErrorValue extends object.DescribedMixin(ErrorValue) {
         return this.describeProperty('message', this.describeAny(this.message))
       }

+      if (emitCause) {
+        emitCause = false
+        return this.describeProperty('cause', this.describeAny(this.cause))
+      }
+
       done = true
       return null
     }
